lib/modules/navigationMiddleware.js

```diff
@@ -118,6 +119,39 @@ export default {
 
           return ret;
         }
+
+        case actions.REDIRECT: {
+          const { url } = action;
+
+          // We want to redirect the current page somewhere else.
+          // If we're on the server, this should always translate into a SET_PAGE
+          // action, because we should issue a proper 3XX status code to redirect.
+          if (isServer) {
+            return store.dispatch(actions.setPage(url));
+          }
+
+          if (url.startsWith('/')) {
+            // We have special logic for relative urls:
+            // Before we route, we want to make sure the app's router
+            // supports the path. It's easy to imagine getting a relative url
+            // that isn't supported, if say there are multiple frontends serving an app
+            // with a load-balancer or router in front that points to the right frontend
+            // based on the request path.
+            const path = url.split('?')[0];
+            const route = matchRoute(path, routes);
+
+            if (route) {
+              const queryParams = extractQuery(url);
+              return store.dispatch(actions.navigateToUrl(METHODS.GET, path, { queryParams }));
+            }
+          }
+
+          // base case for client -- hard redirect via window.location.
+          window.location = url;
+
+          return; // satisfies the no-fallthrough rule for switch-case statements
```

Comment:
I'm fine with that, just for completeness sake? I thought the execution would stop have setting window.location though?